src/init.rs

```diff
@@ -58,3 +51,62 @@ pub fn init(shell_name: &str) {
         print!("{}", script);
     }
 }
+
+/* Bash does not currently support command durations (see issue #124) for details
+https://github.com/starship/starship/issues/124
+*/
+
+const BASH_INIT: &str = r##"
+starship_precmd() {
+        PS1="$(starship prompt --status=$?)";
+};
+PROMPT_COMMAND=starship_precmd;
+"##;
+/* TODO: Once warning/error system is implemented in starship, print a warning
+if starship will not be printing timing due to DEBUG clobber error */
+
+/* For zsh: preexec_functions and precmd_functions provide preexec/precmd in a
+   way that lets us avoid clobbering them.
+
+   Zsh quirk: preexec() is only fired if a command is actually run (unlike in
+   bash, where spamming empty commands still triggers DEBUG). This means a user
+   spamming ENTER at an empty command line will see increasing runtime (since
+   preexec never actually fires to reset the start time).
+
+   To fix this, only pass the time if STARSHIP_START_TIME is defined, and unset
+   it after passing the time, so that we only measure actual commands.
+*/
+
+const ZSH_INIT: &str = r##"
+starship_precmd() {
+    STATUS=$?;
+    if [[ $STARSHIP_START_TIME ]]; then
+        STARSHIP_END_TIME="$(date +%s)";
+        STARSHIP_DURATION=$((STARSHIP_END_TIME - STARSHIP_START_TIME));
+        PROMPT="$(starship prompt --status=$STATUS --cmd-duration=$STARSHIP_DURATION)";
+        unset STARSHIP_START_TIME;
+    else
+        PROMPT="$(starship prompt --status=$STATUS)";
+    fi
+};
+starship_preexec(){
+    STARSHIP_START_TIME="$(date +%s)"
+};
+if [[ ${precmd_functions[(ie)starship_precmd]} -gt ${#precmd_functions} ]]; then
+    precmd_functions+=(starship_precmd);
+fi;
+if [[ ${preexec_functions[(ie)starship_preexec]} -gt ${#preexec_functions} ]]; then
+    preexec_functions+=(starship_preexec);
+fi;
+STARSHIP_START_TIME="$(date +%s)";
+"##;
+
+/* Fish setup is simple because they give us CMD_DURATION. Just account for name
+changes between 2.7/3.0 and do some math to convert ms->s and we can use it */
+const FISH_INIT: &str = r##"
+function fish_prompt;
+    set -l CMD_DURATION "$CMD_DURATION$cmd_duration";
+    set -l starship_duration (math --scale=0 "$CMD_DURATION / 1000");
+    starship prompt --status=$status --cmd-duration=$starship_duration;
```

Comment:
Unfortunately, it looks like a bug snuck in here. 🐛 
Because we are running the `set` commands before the prompt, we lose the old `$status` to the one set by `set`.

It seems we may need to set a variable at the start of the init script to hold onto the `$status` at the time:
```fish
set -l exit_code $status
```

Reply:
Funnily enough, I fixed that for bash and zsh. I'm not sure whether I missed the fact that it was occurring for fish or just forgot to make the edits.

Is there a way we could do automated tests in an actual shell to catch these sort of regressions? I was kind of doing them manually, but even with the basic combinations of `true`/`false` and `sleep` and multiple shells, it was getting a little difficult to figure out if the behavior I was seeing was correct.

Reply:
Agreed. We should introduce full end-to-end tests to catch these bugs. 👍